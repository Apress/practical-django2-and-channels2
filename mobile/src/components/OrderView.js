import React from 'react'
import {
  StyleSheet,
  Text,
  Image,
  TouchableHighlight,
  TouchableOpacity,
  View
} from 'react-native'

import ChatView from './ChatView'

const styles = StyleSheet.create({
  container: {
    paddingTop: 25,
    flex: 1
  },
  orderContainer: {
    borderColor: 'blue',
    borderWidth: 1,
    height: 200
  },
  orderRowOut: {
    borderColor: 'yellow',
    borderWidth: 1,
    height: 52
  },
  orderRowIn: {
    flex: 1,
    flexDirection: 'row',
    alignSelf: 'stretch'
  },
  orderSelected: {
    backgroundColor: 'lightgrey'
  },
  orderImage: {
    borderColor: 'green',
    borderWidth: 1,
    width: 50,
    height: 50
  },
  orderSummary: {
    borderColor: 'red',
    borderWidth: 1,
    flex: 1,
    alignSelf: 'stretch'
  },
  orderPrice: {
    borderColor: 'blue',
    borderWidth: 1
  }
})

function OrderImage (props) {
  if (props.image) {
    return (
      <Image
        style={{ width: 50, height: 50 }}
        source={{ uri: props.backendApi.create_abs_url(props.image) }}
      />
    )
  } else {
    return <View />
  }
}

function OrderTouchArea (props) {
  return (
    <View
      style={[
        styles.orderRowIn,
        props.order.id == props.orderCurrentId ? styles.orderSelected : null
      ]}
    >
      <View style={styles.orderImage}>
        <OrderImage backendApi={props.backendApi} image={props.order.image} />
      </View>
      <Text style={styles.orderSummary}>{props.order.summary}</Text>
      <Text style={styles.orderPrice}>{props.order.price}</Text>
    </View>
  )
}

function OrderSingleView (props) {
  return (
    <TouchableHighlight
      style={styles.orderRowOut}
      onPress={() => props.setOrderId(props.order.id)}
    >
      <OrderTouchArea
        backendApi={props.backendApi}
        order={props.order}
        orderCurrentId={props.orderCurrentId}
      />
    </TouchableHighlight>
  )
}

export default class OrderView extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      orders: [],
      orderCurrentId: null
    }
  }

  componentDidMount () {
    this.props.backendApi
      .fetchOrders()
      .then(orders => this.setOrders(orders))
      .catch(() => alert('Error fetching orders'))
  }

  setOrders (orders) {
    this.setState({
      orders: orders
    })
  }

  render () {
    return (
      <View style={styles.container}>
        <View style={styles.orderContainer}>
          <Text>Your BookTime orders</Text>
          {this.state.orders.map(m => (
            <OrderSingleView
              backendApi={this.props.backendApi}
              key={m.id}
              order={m}
              orderCurrentId={this.state.orderCurrentId}
              setOrderId={orderId => this.setState({ orderCurrentId: orderId })}
            />
          ))}
        </View>
        <ChatView
          backendApi={this.props.backendApi}
          orderCurrentId={this.state.orderCurrentId}
        />
      </View>
    )
  }
}
