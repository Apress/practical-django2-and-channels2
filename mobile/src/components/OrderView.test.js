import React from 'react'
import OrderView from './OrderView'

import renderer from 'react-test-renderer'

it('renders without crashing', () => {
  const mockOrders = [{ id: 5, image: null, summary: '2xaa', price: 22 }]
  const backendApi = {
    fetchOrders: jest.fn(
      () => new Promise((resolve, reject) => resolve(mockOrders))
    ),
    createAbsUrl: jest.fn(url => 'http://booktime.domain' + url)
  }
  const rendered = renderer.create(<OrderView backendApi={backendApi} />)
  expect(rendered).toBeTruthy()
  expect(backendApi.fetchOrders.mock.calls.length).toBe(1)
  rendered.getInstance().setOrders(mockOrders)
  expect(rendered.getInstance().state.orders).toBe(mockOrders)
  expect(rendered.toJSON()).toMatchSnapshot()
})
