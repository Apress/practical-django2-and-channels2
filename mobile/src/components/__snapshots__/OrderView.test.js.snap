// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 25,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "blue",
        "borderWidth": 1,
        "height": 200,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Your BookTime orders
    </Text>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "yellow",
          "borderWidth": 1,
          "height": 52,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "flex": 1,
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          style={
            Object {
              "borderColor": "green",
              "borderWidth": 1,
              "height": 50,
              "width": 50,
            }
          }
        >
          <View />
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "stretch",
              "borderColor": "red",
              "borderWidth": 1,
              "flex": 1,
            }
          }
        >
          2xaa
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "borderColor": "blue",
              "borderWidth": 1,
            }
          }
        >
          22
        </Text>
      </View>
    </View>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      No order selected for chat
    </Text>
  </View>
</View>
`;
